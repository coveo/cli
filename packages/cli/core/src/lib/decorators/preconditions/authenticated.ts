import {
  PreconditionError,
  PreconditionErrorCategory,
} from '../../errors/preconditionError';
import {
  AuthenticationStatus,
  getAuthenticationStatus,
} from '../../platform/authenticatedClient';

export function IsAuthenticated() {
  return async function () {
    const status = await getAuthenticationStatus();
    if (status === AuthenticationStatus.LOGGED_OUT) {
      throw new PreconditionError(
        'Not currently logged in. Run coveo auth:login first.',
        {category: PreconditionErrorCategory.Authentication}
      );
    }

    if (status === AuthenticationStatus.EXPIRED) {
      throw new PreconditionError(
        'Authentication token is expired. Run coveo auth:login first.',
        {category: PreconditionErrorCategory.Authentication}
      );
    }
  };
}
