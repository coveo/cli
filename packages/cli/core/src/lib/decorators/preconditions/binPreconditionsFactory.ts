import type {Command} from '@oclif/core';
import {dedent} from 'ts-dedent';
import {spawnProcessOutput, SpawnProcessOutput} from '../../utils/process';
import {satisfies, validRange} from 'semver';
import {
  PreconditionError,
  PreconditionErrorCategory,
} from '../../errors/preconditionError';

export interface BinPreconditionsOptions {
  params?: Array<string>;
  prettyName: string;
  installLink?: string;
  howToInstallBinText?: string;
}

const defaultOptions: Required<Omit<BinPreconditionsOptions, 'prettyName'>> = {
  params: ['--version'],
  installLink:
    'https://github.com/coveo/cli/wiki/Node.js,-NPM-and-NPX-requirements',
  howToInstallBinText: '',
};

export function getBinVersionPrecondition(
  binaryName: string,
  options: BinPreconditionsOptions,
  versionExtractor: (stdout: string) => string = (stdout) => stdout
) {
  const appliedOptions: Required<BinPreconditionsOptions> = {
    ...defaultOptions,
    ...options,
    prettyName: options.prettyName,
  };
  return function (versionRange: string) {
    return async function (target: Command) {
      if (!validRange(versionRange)) {
        const message = dedent`
          Required version invalid: "${versionRange}".
          Please report this error to Coveo: https://github.com/coveo/cli/issues/new
        `;
        throw new PreconditionError(message, {
          category: PreconditionErrorCategory.InvalidBinVersionRange,
        });
      }

      const output = await spawnProcessOutput(
        binaryName,
        appliedOptions.params
      );

      await checkIfBinIsInstalled(target, binaryName, appliedOptions, output);
      const version = versionExtractor(output.stdout);

      if (!satisfies(version, versionRange)) {
        const message = dedent`
          ${target.id} needs a ${
          appliedOptions.prettyName
        } version in this range: "${versionRange}"
          Version detected: ${version}

          ${warnHowToInstallBin(appliedOptions)}
        `;
        throw new PreconditionError(message, {
          category: PreconditionErrorCategory.InvalidBinVersionRange,
        });
      }
    };
  };
}

export function getBinInstalledPrecondition(
  binaryName: string,
  options: BinPreconditionsOptions
) {
  const appliedOptions: Required<BinPreconditionsOptions> = {
    ...defaultOptions,
    ...options,
    prettyName: options.prettyName,
  };
  return function () {
    return async function (target: Command) {
      const output = await spawnProcessOutput(
        binaryName,
        appliedOptions.params
      );
      await checkIfBinIsInstalled(target, binaryName, appliedOptions, output);
    };
  };
}

async function checkIfBinIsInstalled(
  target: Command,
  binaryName: string,
  options: Required<BinPreconditionsOptions>,
  output: SpawnProcessOutput
): Promise<void | never> {
  if (output.exitCode === 'ENOENT') {
    const warningMessage = dedent`${target.id} requires ${
      options.prettyName
    } to run.

    ${warnHowToInstallBin(options)}`;
    throw new PreconditionError(warningMessage, {
      category: PreconditionErrorCategory.MissingBin,
    });
  }

  if (output.exitCode !== '0') {
    const message = dedent`
      ${target.id} requires a valid ${options.prettyName} installation to run.
      An unknown error happened while running ${binaryName} ${options.params.join(
      ' '
    )}.
      ${output.stderr}

      ${warnHowToInstallBin(options)}
    `;
    throw new PreconditionError(message, {
      category: PreconditionErrorCategory.InvalidBinInstallation,
    });
  }
}

function warnHowToInstallBin(options: BinPreconditionsOptions) {
  const newParagraph = '\n\n';
  const howToInstallMessage = [];
  if (options.howToInstallBinText) {
    howToInstallMessage.push(options.howToInstallBinText);
  }
  howToInstallMessage.push(
    `Please visit ${options.installLink} for more detailed installation information.`
  );
  return howToInstallMessage.join(newParagraph);
}
