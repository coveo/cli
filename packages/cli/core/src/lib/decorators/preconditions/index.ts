// TODO: Remove, unusued
export * from './git';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './node';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npx';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './npm';
// TODO: Move to `@coveo/plugin-cli-ui
export * from './ng';
