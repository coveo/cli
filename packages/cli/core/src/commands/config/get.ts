import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {Config} from '@coveo/cli-commons/config/config';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Get extends CLICommand {
  public static description = 'Display the current Coveo CLI configuration.';

  public static args = [
    {
      name: 'key',
      description: 'The config key for which to show the value',
      required: false,
    },
  ];

  public static examples: Example[] = [
    {
      command: 'coveo config:get',
      description: 'Get all the configuration values',
    },
    {
      command: 'coveo config:get organization',
      description: 'Get the organization to which you are connected',
    },
    {
      command: 'coveo config:get accessToken',
      description: 'Get the access token given to you by the Coveo Platform',
    },
  ];

  @Trackable()
  public async run() {
    const {args} = await this.parse(Get);
    const cfg = new Config(this.config.configDir);
    const keysToRender = args.key ? [args.key] : undefined;
    ConfigRenderer.render(cfg, keysToRender);
  }
}
