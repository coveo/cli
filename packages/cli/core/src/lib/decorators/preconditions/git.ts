import {getBinInstalledPrecondition} from './binPreconditionsFactory';

export const IsGitInstalled = getBinInstalledPrecondition('git', {
  prettyName: 'Git',
  installLink: 'https://git-scm.com/book/en/v2/Getting-Started-Installing-Git',
});
