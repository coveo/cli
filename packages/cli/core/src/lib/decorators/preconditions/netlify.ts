import {appendCmdIfWindows} from '../../utils/os';
import {getBinVersionPrecondition} from './binPreconditionsFactory';

const angularCliVersionMatcher = /^netlify-cli\/(?<version>\d+\.\d+\.\d+).*$/m;

export const IsNetlifyCliVersionInRange = getBinVersionPrecondition(
  appendCmdIfWindows`netlify`,
  {
    prettyName: 'Netlify-CLI',
    howToInstallBinText:
      'You can install the Netlify-CLI by running npm i -g netlify-cli',
    installLink: 'https://docs.netlify.com/cli/get-started/#installation',
    params: ['version'],
  },
  (stdout: string) =>
    stdout.match(angularCliVersionMatcher)?.groups?.['version'] ?? ''
);
