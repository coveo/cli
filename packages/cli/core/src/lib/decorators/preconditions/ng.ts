import {appendCmdIfWindows} from '@coveo/cli-commons/utils/os';
import {getBinVersionPrecondition} from './binPreconditionsFactory';

const angularCliVersionMatcher = /^Angular CLI: (?<version>\d+\.\d+\.\d+)$/m;

export const IsNgVersionInRange = getBinVersionPrecondition(
  appendCmdIfWindows`ng`,
  {
    prettyName: 'Angular-CLI',
    howToInstallBinText:
      'You can install the Angular-CLI by running npm i -g @angular/cli',
    installLink: 'https://angular.io/guide/setup-local#install-the-angular-cli',
    params: ['version'],
  },
  (stdout: string) =>
    stdout.match(angularCliVersionMatcher)?.groups?.['version'] ?? ''
);
