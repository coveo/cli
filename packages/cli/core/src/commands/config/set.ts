import {CLICommand} from '@coveo/cli-commons/command/cliCommand';
import {CliUx, Flags} from '@oclif/core';
import {Config} from '@coveo/cli-commons/config/config';
import {AuthenticatedClient} from '@coveo/cli-commons/platform/authenticatedClient';
import {
  IsAuthenticated,
  Preconditions,
} from '@coveo/cli-commons/preconditions/index';
import {Trackable} from '@coveo/cli-commons/preconditions/trackable';
import {InvalidCommandError} from '../../lib/errors/InvalidCommandError';
import {ConfigRenderer} from '@coveo/cli-commons/config/configRenderer';
import type {Example} from '@oclif/core/lib/interfaces';

export default class Set extends CLICommand {
  public static description = 'Modify the current Coveo CLI configuration.';

  public static flags = {
    // TODO CDX-1246
    environment: Flags.string({
      char: 'e',
      hidden: true,
    }),
    // TODO CDX-1246
    region: Flags.string({
      char: 'r',
      hidden: true,
    }),
    organization: Flags.string({
      char: 'o',
      description:
        'The identifier of the organization inside which to perform operations. See <https://docs.coveo.com/en/1562/#organization-id-and-other-information>.',
      helpValue: 'myOrgID',
    }),
  };

  public static examples: Example[] = [
    {
      command: 'coveo config:set --organization myOrgId',
      description: 'connect to a different organization',
    },
  ];

  @Trackable()
  @Preconditions(IsAuthenticated())
  public async run() {
    const {flags} = await this.parse(Set);
    // TODO CDX-1246
    if (flags.environment || flags.region) {
      CliUx.ux.error(
        'To connect to a different region or environment, use the `auth:login` command'
      );
    }
    if (Object.entries(flags).length === 0) {
      throw new InvalidCommandError('Command should contain at least 1 flag');
    }
    const cfg = new Config(this.config.configDir);
    if (flags.organization) {
      await this.verifyOrganization(flags.organization);
      cfg.set('organization', flags.organization);
    }

    ConfigRenderer.render(cfg);
  }

  private async verifyOrganization(org: string) {
    const hasAccess = await new AuthenticatedClient().getUserHasAccessToOrg(
      org
    );
    if (!hasAccess) {
      this.error(
        `You either don't have access to organization ${org}, or it doesn't exist.`
      );
    }
  }
}
