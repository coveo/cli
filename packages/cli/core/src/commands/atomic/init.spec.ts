import {test} from '@oclif/test';

jest.mock('inquirer');
import inquirer from 'inquirer';
jest.mock('@coveo/cli-commons/config/config');
import {Config} from '@coveo/cli-commons/config/config';
jest.mock('../../lib/atomic/createAtomicProject');
import {
  createAtomicApp,
  createAtomicLib,
} from '../../lib/atomic/createAtomicProject';
jest.mock('@coveo/cli-commons/preconditions');
import {
  HasNecessaryCoveoPrivileges,
  IsAuthenticated,
} from '@coveo/cli-commons/preconditions';
jest.mock('../../lib/decorators/preconditions');
import {
  IsNpxInstalled,
  IsNodeVersionInRange,
} from '../../lib/decorators/preconditions';

import {mockPreconditions} from '@coveo/cli-commons/preconditions/mockPreconditions';

describe('atomic:init', () => {
  const mockedCreateAtomicApp = jest.mocked(createAtomicApp);
  const mockedCreateAtomicLib = jest.mocked(createAtomicLib);
  const mockedInquirer = jest.mocked(inquirer);
  const mockedConfig = jest.mocked(Config);
  const mockedConfigGet = jest.fn();
  const mockedIsNpxInstalled = jest.mocked(IsNpxInstalled);
  const mockedIsNodeVersionInRange = jest.mocked(IsNodeVersionInRange);
  const mockedApiKeyPrivilege = jest.mocked(HasNecessaryCoveoPrivileges);
  const mockedIsAuthenticated = jest.mocked(IsAuthenticated);
  const doMockConfig = () => {
    mockedConfigGet.mockReturnValue({
      region: 'us',
      organization: 'default-org',
      environment: 'prod',
    });

    mockedConfig.mockImplementation(
      () =>
        ({
          get: mockedConfigGet,
        } as unknown as Config)
    );
  };

  const preconditionStatus = {
    node: true,
    npx: true,
    apiKey: true,
    authentication: true,
  };

  const doMockPreconditions = function () {
    const mockedPreconditions = mockPreconditions(preconditionStatus);
    mockedIsNodeVersionInRange.mockReturnValue(mockedPreconditions.node);
    mockedIsNpxInstalled.mockReturnValue(mockedPreconditions.npx);
    mockedApiKeyPrivilege.mockReturnValue(mockedPreconditions.apiKey);
    mockedIsAuthenticated.mockReturnValue(mockedPreconditions.authentication);
  };

  beforeEach(() => {
    jest.resetAllMocks();
    doMockConfig();
    doMockPreconditions();
  });

  test
    .stdout()
    .stderr()
    .command(['atomic:init', '--type=lib', 'foo'])
    .it('calls `createAtomicLib` properly', () => {
      expect(mockedCreateAtomicLib).toBeCalledTimes(1);
      expect(mockedCreateAtomicLib.mock.lastCall).toMatchSnapshot();
    });

  test
    .stdout()
    .stderr()
    .command(['atomic:init', '--type=library', 'foo'])
    .it('calls `createAtomicLib` properly', () => {
      expect(mockedCreateAtomicLib).toBeCalledTimes(1);
      expect(mockedCreateAtomicLib.mock.lastCall).toMatchSnapshot();
    });

  test
    .stdout()
    .stderr()
    .command(['atomic:init', '--type=app', 'foo'])
    .it('calls `createAtomicApp` properly', () => {
      expect(mockedCreateAtomicApp).toBeCalledTimes(1);
      expect(mockedCreateAtomicApp.mock.lastCall).toMatchSnapshot();
    });

  test
    .stdout()
    .stderr()
    .command(['atomic:init', '--type=application', 'foo'])
    .it('calls `createAtomicApp` properly', () => {
      expect(mockedCreateAtomicApp).toBeCalledTimes(1);
      expect(mockedCreateAtomicApp.mock.lastCall).toMatchSnapshot();
    });

  test
    .do(() => {
      mockedInquirer.prompt.mockResolvedValue({type: 'application'});
    })
    .stdout()
    .stderr()
    .command(['atomic:init', 'foo'])
    .it(
      'calls `createAtomicApp` when the user select application in the prompt',
      () => {
        expect(mockedCreateAtomicApp).toBeCalledTimes(1);
        expect(mockedCreateAtomicApp.mock.lastCall).toMatchSnapshot();
      }
    );

  test
    .do(() => {
      mockedInquirer.prompt.mockResolvedValue({type: 'library'});
    })
    .stdout()
    .stderr()
    .command(['atomic:init', 'foo'])
    .it(
      'calls `createAtomicLib` when the user select library in the prompt',
      () => {
        expect(mockedCreateAtomicLib).toBeCalledTimes(1);
        expect(mockedCreateAtomicLib.mock.lastCall).toMatchSnapshot();
      }
    );
});
