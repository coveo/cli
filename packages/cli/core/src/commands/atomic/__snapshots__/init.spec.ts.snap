// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`atomic:init calls \`createAtomicApp\` properly 1`] = `
[
  {
    "cfg": {
      "environment": "prod",
      "organization": "default-org",
      "region": "us",
    },
    "projectName": "foo",
  },
]
`;

exports[`atomic:init calls \`createAtomicApp\` properly 2`] = `
[
  {
    "cfg": {
      "environment": "prod",
      "organization": "default-org",
      "region": "us",
    },
    "projectName": "foo",
  },
]
`;

exports[`atomic:init calls \`createAtomicApp\` when the user select application in the prompt 1`] = `
[
  {
    "cfg": {
      "environment": "prod",
      "organization": "default-org",
      "region": "us",
    },
    "projectName": "foo",
  },
]
`;

exports[`atomic:init calls \`createAtomicLib\` properly 1`] = `
[
  {
    "projectName": "foo",
  },
]
`;

exports[`atomic:init calls \`createAtomicLib\` properly 2`] = `
[
  {
    "projectName": "foo",
  },
]
`;

exports[`atomic:init calls \`createAtomicLib\` when the user select library in the prompt 1`] = `
[
  {
    "projectName": "foo",
  },
]
`;
